polar-core/src/partial/isa_constraint_check.rs

```diff
@@ -25,20 +26,31 @@ impl IsaConstraintCheck {
             existing,
             proposed_tag,
             result: None,
+            alternative_check: None,
             last_call_id: 0,
         }
     }
 
-    /// Check if the existing constraints set is compatible with the proposed
-    /// matches class.
+    /// Check if existing constraints are compatible with the proposed type constraint (`_this
+    /// matches Post{}`).
     ///
-    /// Returns: None if compatible, QueryEvent::Done { false } if incompatible,
-    /// or QueryEvent to ask for compatibility.
+    /// If the existing constraint is also a type constraint , we return a pair of
+    /// `QueryEvent::ExternalIsSubclass`es to check whether the type constraints are compatible.
+    /// The constraints are compatible if they are the same class or if either is a subclass of the
+    /// other.
+    ///
+    /// If the existing constraint is not a type constraint, there's no external check required,
+    /// and we return `None` to indicate compatibility.
+    ///
+    /// Returns:
+    /// - `None` if compatible.
+    /// - A pair of `QueryEvent::ExternalIsSubclass` checks if compatibility cannot be determined
+    /// locally.
```

Comment:
One other case that this doesn't account for, which comes up in languages that support multiple inheritance:

- If you have these three classes:
```python
class Doctor:
    pass

class Administrator:
    pass

class DepartmentHead(Doctor, Administrator):
    pass
```

And then rules that looked like this:

```python
f(x: Doctor);
g(x: Administrator);
```
Those constraints would be incompatible by this implementation, although there exists a class (`DepartmentHead`) for which they are compatible. 

I don't think that this is a common use case AT ALL, but we should probably document the behavior and say that we won't recognize multiple inheritance structures with Partials?

Reply:
I don't quite follow -- what's the actual rule that's incompatible under the current implementation?

Reply:
Updated my comment to try to be more clear

Reply:
I may be misunderstanding how the engine evaluates it though. If it only compares each `isa` constraint to the initial `Partial` type constraint, and not to each other, than it would be fine.

Reply:
The two `matches` constraints would exist on separate branches of an `or`. The partial's constraints would look something like this:
```python
_this matches DepartmentHead{}
and ((_this matches Doctor{} and _this.approver = true)
or (_this matches Administrator{} and _this.approver = true))
```
I added parens to hopefully make it clearer.

Reply:
Ah yeah, I meant for them to be on the same `and` branch, my bad. E.g., 

```
is_department_head(x: DepartmentHead) if is_doctor(x) and is_administrator(x);
is_doctor(x: Doctor);
is_administrator(x: Administrator);
```

Wondering if this would cause the constraints to clash or not.

Reply:
Ah, you're right! We'll add a failing test.

Right now we would get to the point of adding the `_this matches Administrator{}` constraint, check it against the existing `_this matches DepartmentHead{}` constraint (which would succeed), and then check it against the existing `_this matches Doctor{}` constraint, which _should_ succeed but will currently fail.

Reply:
(This PR is still a step in the direction of correctness, though)

Reply:
Thanks for checking that! Yeah again, I think this is a very niche case, especially since most of our mainstream use cases don't seem to use complex multi-dispatch.